job-info: Support new "exception" list attributes

chu11 opened the pull request:
Build on top of PR #2831, support 4 new job listing attributes, "exception", "exception_severity", "exception_type", and "exception_note".  Hopefully what these are don't require explanation.

If an exception occurred, return true/false for `exception`.  if true, the others will contain the values for the highest severity and the type/note that went along with it.

```
>flux jobs -A --format="{exception},{exception_severity},{exception_type},{exception_note}"
EXCEPTION,EXCEPTION-SEVERITY,EXCEPTION-TYPE,EXCEPTION-NOTE
True,0,exec,task 0: start failed: fdasfdasf: No such file or directory
```

I'm not the most excited about the long format names and the headers, but just put those in for place holders.  Wasn't sure of the best abbreviation ... google says abbreviation is "exc", but my gut says to use "excp"?  should headers just be severity, type, & note?

SteVwonder commented on doc/man1/flux-jobs.adoc:
```diff
@@ -81,6 +81,16 @@ nnodes:: job node count (if job ran / is running), empty string otherwise
+exception:: True of False if job had an exception, empty string otherwise
+exception_hyphen:: same as exception, but '-' instead of empty string
+exception_severity:: If exception True, the highest severity, empty string otherwise
+exception_severity_hype:: same as exception_severity, but '-' instead of empty string
```
LET'S GET HYPE!!!!!

Totally subjective but I think just the straight truncation (`exception_severity_hyph`) might be better here.

SteVwonder replied:
Another random idea: rather than having a plethora of `_hyphen` attributes, I wonder if we could express this in the `format_spec` part of the format string.  For example `{exception_severity:h}` where `h` invokes a [custom conversion type](https://stackoverflow.com/questions/19864302/add-custom-conversion-types-for-string-formatting).

chu11 replied:
> Another random idea: rather than having a plethora of _hyphen attributes, I wonder if we could express this in the format_spec part of the format string. For example {exception_severity:h} where h invokes a custom conversion type.

This is a great idea.  Not for this PR, wrote up issue #2859 for it

SteVwonder commented on doc/man1/flux-jobs.adoc:
```diff
@@ -81,6 +81,16 @@ nnodes:: job node count (if job ran / is running), empty string otherwise
+success_hyphen:: same as success, but '-' if job has not yet completed
+exception:: True of False if job had an exception, empty string otherwise
+exception_hyphen:: same as exception, but '-' instead of empty string
+exception_severity:: If exception True, the highest severity, empty string otherwise
```
Random idea, but maybe using `.` for the objects nested within exception will feel more natural to users familar with OOP.
```suggestion
exception.severity:: If exception True, the highest severity, empty string otherwise
```

chu11 replied:
I like the idea, but I don't think it'll work.  If a string format field is done like this in python, i.e. `{exception.severity}`, won't Python think `exception` is an object and severity is a field within it?

My test:

```
class A:
    def __init__(self):
        self.a = "a"
        self.b = "b"

class Foo:
    def __init__(self):
        self.A = A()


str="{0.A.b}"
o = Foo()
print(str.format(o))
```

I'm sure there are tricks to get around this, but dunno if its worth the energy?

grondo replied:
> If a string format field is done like this in python, i.e. {exception.severity}, won't Python think exception is an object and severity is a field within it?

I assumed that is what @SteVwonder was suggesting. Create an `ExceptionInfo` object and make `severity`, `type`, etc. attributes of that object. This is way is more flexible than the generating the underscore names by hand...

SteVwonder replied:
> Create an ExceptionInfo object and make severity, type, etc. attributes of that object. This is way is more flexible than the generating the underscore names by hand...

Yeah, I like @grondo's suggestion.  TBH, I didn't realize that python interpreted the `.` in format strings and called `getattr`, so I originally was just suggesting a string literal change.  But even so, I think the `.` syntax has promise for flexibility and extensibility.

chu11 replied:
> I assumed that is what @SteVwonder was suggesting. Create an ExceptionInfo object and make severity, type, etc. attributes of that object. This is way is more flexible than the generating the underscore names by hand...

Oh! I just thought he meant replacing the `_` with `.` naming wise.  Lemme look into this idea.

chu11 replied:
Hmmm.  Fiddling with this tonight.  Right now the attribute `exception` returns true or false.   i.e. `JobInfoObject.exception` is a boolean.  If we make `exception` an object, Would users have to specify `exception.exception` (or some other word) to get the boolean value?  I know we can get the string representation by overriding `__repr__` and/or  `__str__`, but that's not ideal.

SteVwonder replied:
>  I know we can get the string representation by overriding __repr__ and/or __str__

The JobException class would be local to `flux-jobs`, right?  If so, overriding `__str__` seems like a viable option to me.  Were there any particular downsides to that override that you were thinking about/trying to avoid?

> Would we want exception.exception as the ideal format option?

Maybe something like `exception.occurred`?

grondo replied:
> Maybe something like exception.occurred?

I agree, though I think eventually the JobInfo and ExceptionInfo classes might be useful outside of `flux-jobs`.

Rather than being useful as an output format field (I don't see a real use case for printing `True`/`False` in the output), it seems like `exception.type` or `exception.severity` would be the most useful in `flux-jobs` output. Something like `exception.occurred` would be more useful as a filtering option down the road.

I guess one thing I didn't think of is that every job will need to have an ExceptionInfo object instantiated even if it means there was no exception, o/w something like `{exception.occurred}` would throw some kind of ... exception??

That is a downside to this approach I guess, though I bet with python there's a clever way to handle it.

chu11 replied:
> Were there any particular downsides to that override that you were thinking about/trying to avoid?

Well at some point (such as for a derived state), we'd want to actually do boolean checks on an exception to determine what derived state to output.  i.e. `JobInfo.exception == True`, which I don't think overriding `__str__` or `__repr__` would work.

> I guess one thing I didn't think of is that every job will need to have an ExceptionInfo object instantiated even if it means there was no exception, o/w something like {exception.occurred} would throw some kind of ... exception??

Yeah, I think so.  In addition, since `JobInfo` is initialed like this:

```
        #  Set defaults, then update with job-info.list response items:
        combined_dict = self.defaults.copy()
        combined_dict.update(info_resp)
```

We have to override `exception` if it was in the list response.

grondo replied:
> Well at some point (such as for a derived state), we'd want to actually do boolean checks on an exception to determine what derived state to output. i.e. JobInfo.exception == True, which I don't think overriding __str__ or __repr__ would work.

Wouldn't you check `job.exception.occurred`?

chu11 replied:
> Wouldn't you check job.exception.occurred?

Now we would, yes.  I was referring to pre-ExceptionInfo class.  Sorry if that was confusing :P

grondo replied:
got it, sorry! :+1:

dongahn commented:
@chu11: speaking of making use of a "note", is flux-sched issue https://github.com/flux-framework/flux-sched/issues/626 relevant?

chu11 commented:
@dongahn in this particular case no, since we're since we're just grabbing the "note" provided by the exception event in the eventlog.  I believe the "note" in your other issue is sort of a special 1-off note that has to be grabbed in a different way.

chu11 commented:
rebased and re-pushed now that #2865 is in.

chu11 commented:
rebased, fixed up issue after #2850 merged in.

Added a `ExceptionInfo` so format specifiers are now `{exception.occurred}`, `{exception.type}`, etc.

I really hate this code:

```
+        # Overwrite "exception" with our exception object
+        e1 = combined_dict["exception_occurred"] if "exception_occurred" in combined_dict else ""
+        e2 = combined_dict["exception_severity"] if "exception_severity" in combined_dict else ""
+        e3 = combined_dict["exception_type"] if "exception_type" in combined_dict else ""
+        e4 = combined_dict["exception_note"] if "exception_note" in combined_dict else ""
+        combined_dict["exception"] = ExceptionInfo(e1, e2, e3, e4)
```

and

```
+        "exception.occurred": "EXCEPTION-OCCURRED",
+        "exception.severity": "EXCEPTION-SEVERITY",
+        "exception.type": "EXCEPTION-TYPE",
+        "exception.note": "EXCEPTION-NOTE",
```

Its so long, I really need to come up with an abbreviation.

But in principle, got things working as we discussed above.

SteVwonder commented:
> I really hate this code:

One thought, having `combined_dict` be a defaultdict might help with all the `if "exception_*" in combined_dict else ""` code: https://docs.python.org/3.7/library/collections.html#collections.defaultdict

chu11 commented:
> One thought, having combined_dict be a defaultdict might help with all the if "exception_*" in combined_dict else "" code: https://docs.python.org/3.7/library/collections.html#collections.defaultdict

I just realized, I could just use `combined_dict.get()` :-) re-pushed

garlick commented:
Is this just waiting on a rebase to go in?

chu11 commented:
> Is this just waiting on a rebase to go in?

yup, now that #2831 and #2886 gotta fix this up.

chu11 commented:
rebased and re-pushed, lgtm had some issues but seems to have passed for now

grondo commented:
Any idea what I'm doing wrong? The tests seem to pass, but a basic check on the cmdline gives me this error:
```
[user1@1c2d69ee7a8b flux-core]$ FLUX_HANDLE_TRACE=1 FLUX_PYCLI_LOGLEVEL=10 flux jobs -a --format='{exception.occurred}'
--------------------------------------
>[000] ||
>[013] job-info.list
>[084] {"max_entries": 1000, "attrs": ["exception_occurred"], "userid": 2000, "states": 62}
>[020] 8E01010BFFFFFFFF00000000FFFFFFFF00000001
--------------------------------------
<[000] ||
<[013] job-info.list
<[103] {"jobs":[{"id":247380049920,"exception_occurred":false},{"id":150156083200,"exception_occurred":true}]}
<[020] 8E01020B000007D0000000010000000000000001
flux-jobs: ERROR: 'exception'
flux-jobs: DEBUG: Traceback (most recent call last):
  File "/home/user1/flux-core/src/bindings/python/flux/util.py", line 134, in __call__
    main_func()
  File "/home/user1/flux-core/src/cmd/flux-jobs.py", line 441, in main
    print(formatter.header())
  File "/home/user1/flux-core/src/cmd/flux-jobs.py", line 418, in header
    return self.JobFormatter().format(self.header_format(), **self.headings)
  File "/usr/lib64/python3.6/string.py", line 190, in format
    return self.vformat(format_string, args, kwargs)
  File "/usr/lib64/python3.6/string.py", line 194, in vformat
    result, _ = self._vformat(format_string, args, kwargs, used_args, 2)
  File "/usr/lib64/python3.6/string.py", line 234, in _vformat
    obj, arg_used = self.get_field(field_name, args, kwargs)
  File "/usr/lib64/python3.6/string.py", line 299, in get_field
    obj = self.get_value(first, args, kwargs)
  File "/usr/lib64/python3.6/string.py", line 256, in get_value
    return kwargs[key]
KeyError: 'exception'

```

chu11 commented:
> Any idea what I'm doing wrong

Hmm I clearly screwed up something.  My initial guess is it was related to the header.

```
>flux jobs --suppress-header -a --format="{exception.occurred}"
False
>flux jobs -a --format="{exception.occurred}"
flux-jobs: ERROR: 'exception'
```

Will take a look.  And this indicates I need some new tests for the header too.

grondo commented:
Ah, clever catch!

chu11 commented:
re-pushed with a fix.  I'm not sure how this worked before when I was testing.  I clearly changed something and it stopped working at some point.

So this was a little tricky and the solution I added is a bit hackish.

- I have to create an exception object to pass to `.format()` so that the appropriate fields are output.  No biggie, this part is clear.

- But the `headings` dictionary is used to check if the format fields passed in by the user are legal.  So the `exception.X` fields need to be in there as well.  So I added it in as a special case.

- Maybe there is something Pythonic that could be done other than what I did, but I couldn't figure one out.

- Added new tests to test all format fields for header output.

grondo commented:
For now that seems reasonable. But I bet in the future we could do something more automatic by checking attributes of the class or somesuch (above my Python wizarding level)

grondo commented:
BTW, this was working for me now. Want me to set MWP?

chu11 commented:
I'll set it :-) Thanks!
